File: 50dias/cpp/21.cpp
#include <iostream>
#include <fstream>
#include <list>

using namespace std;

void lerPalavras(list<string>& palavras) {
    ifstream inputFile;
    inputFile.open("input.txt");
    
    string palavra;
    while (inputFile >> palavra)
        palavras.push_back(palavra);
    inputFile.close();
}

void ordenaPalavrasPorTamanhoEmOrdemDescrescente (
    list<string>& palavras
)
{
    auto ordenador = []
    (const string& first, const string& second) {
        return first.size() > second.size();
    };
    palavras.sort(ordenador);
}

void preencheMaioresPalavras(
    list<string>& palavras,
    list<string>& maioresPalavras
)
{
    int tamanhoMaiorPalavra = palavras.front().size();
    for (auto palavra : palavras) {
        if (palavra.size() == tamanhoMaiorPalavra)
            maioresPalavras.push_back(palavra);
    }
}

void escreveMaioresPalavrasNoArquivo(list<string>& palavras) {
    list<string> maioresPalavras;
    preencheMaioresPalavras(palavras, maioresPalavras);
    
    ofstream outputFile;
    outputFile.open("output.txt");
    
    if (maioresPalavras.size() > 1)
        outputFile << "As maiores palavras são: " << endl;
    else
        outputFile << "A maior palavra é: ";
    
    for (auto palavra : maioresPalavras)
        outputFile << palavra << endl;
    
    outputFile.close();
}

int main()
{
    list<string> palavras;
    
    lerPalavras(palavras);
    ordenaPalavrasPorTamanhoEmOrdemDescrescente(palavras);
    
    escreveMaioresPalavrasNoArquivo(palavras);
    
    return 0;
}

File: 50dias/python/21.py

def lerPalavrasDoArquivo(palavras):
    input_file = open('input.txt', 'r')
    conteudo = input_file.readlines()
    input_file.close()
    
    for linha in conteudo:
        palavrasDaLinha = linha.split()
        palavras += palavrasDaLinha
    
def ordenaPalavrasPorTamanhoEmOrdemDescrescente(palavras):
    palavras.sort(key = len, reverse=True)
    
def preencheMaioresPalavras(palavras, maioresPalavras):
    tamanhoMaiorPalavra =  len(palavras[0])
    for palavra in palavras:
        tamanhoPalavra = len(palavra)
        if tamanhoPalavra == tamanhoMaiorPalavra:
            maioresPalavras.append(palavra)
        else:
            break
        
def escreveNoArquivoDeSaida(maioresPalavras):
    mensagem = "\n\n"
    
    multiplasPalavras = len(maioresPalavras) > 1
    if multiplasPalavras:
        mensagem += "As maiores palavras são: "
    else :
        mensagem += "A maior palavra é: "
    
    output_file = open('output.txt', 'w')
    output_file.write(mensagem + '\n')
    
    for palavra in maioresPalavras:
        output_file.write(palavra)
        output_file.write('\n')
    output_file.close()
        
def main():
    palavras = []
    
    lerPalavrasDoArquivo(palavras)

    ordenaPalavrasPorTamanhoEmOrdemDescrescente(palavras)
    
    maioresPalavras = []
    preencheMaioresPalavras(palavras, maioresPalavras)
    
    escreveNoArquivoDeSaida(maioresPalavras)
    
main()